torch/export/dynamic_shapes.py

```diff
@@ -904,6 +909,149 @@ def check_shape(path, t, dynamic_shape):
 
     _tree_map_with_path(check_shape, combined_args, dynamic_shapes, tree_name="inputs")
 
+    # raise user warning if both True & Dims are specified in dynamic_shapes
+    flat_dynamic_shapes = _flatten_dynamic_shapes(combined_args, dynamic_shapes)
+    flatter_dynamic_shapes, _ = tree_flatten(flat_dynamic_shapes)
+    if (
+        any(isinstance(s, _Dim) for s in flatter_dynamic_shapes)
+        and any(s is True for s in flatter_dynamic_shapes)
+    ):
+        raise UserError(
+            UserErrorType.INVALID_INPUT,
+                "Specifying both `True` and `Dim` or `DerivedDim` in `dynamic_shapes` is not well supported at the moment, " \
+                "and can easily lead to constraint violation errors or obscure errors in torch.export. " \
+                "Dim/DerivedDims expect all equal or related dimensions to be specified, and does not yet compose well with `True`. " \
+                "We suggest using `True` mixed with `None` for auto-dynamic + static shapes, plus torch._check(dim >= min), "
+                "torch._check(dim <= max) calls in your program to specify min/max ranges, or `Dim`/`DerivedDim` mixed with `None` " \
+                "if you want to assert on the exact specification of your program's dynamic shapes behavior.",
+                case_name="dynamic_shapes_validation",
+            )
+
+
+def _transform_shapes_for_default_dynamic(
```

Comment:
This entire function is unnecessary if we could ignore BC and completely change `_dynamo.export()` behavior, but this seems safer to start with.

Reply:
I think it is better to make the full spec explicit early in the export pathway anyway, rather than carry the meanings of defaults and interpret them everywhere. Makes it easier to adapt to automated use cases where we can just generate complete specs from side info (in addition to continuing to support user-defined specs, which have some pressure to be minimal).